### PR TITLE
feat(models-http-api): add rate limit in embedding api

### PR DESCRIPTION
**Pull request by zwpaper**
Rate enabled, check the highlighted seconds gap between several requests:

CleanShot 2024-11-22 at 21 25 52@2x

### CODE REVIEW

File: crates/tabby-common/src/config.rs [hunk 1]
```diff
@@ -309,6 +312,14 @@ pub struct HttpModelConfig {
     pub additional_stop_words: Option<Vec<String>>,
 }
 
+#[derive(Serialize, Deserialize, Builder, Debug, Clone)]
+pub struct RequestLimit {
+    // The limited number of requests can be made in following `per` time period.
+    pub num_request: u64,
+    // The time period in seconds to limit the number of requests.
+    pub per: u64,
```
Comment (wsxiaoys) [inline crates/tabby-common/src/config.rs:320]:
```suggestion
    /// ...
    pub request_per_minute: u64
```

File: crates/http-api-bindings/src/lib.rs [hunk 2]
```diff
@@ -5,6 +5,13 @@ mod embedding;
 pub use chat::create as create_chat;
 pub use completion::{build_completion_prompt, create};
 pub use embedding::create as create_embedding;
+use reqwest::Response;
+use serde_json::Value;
+
+#[async_trait::async_trait]
+pub trait HttpClient: Send + Sync {
+    async fn post(&self, url: &str, body: Value) -> anyhow::Result<Response>;
+}
```
Comment (wsxiaoys) [inline crates/http-api-bindings/src/lib.rs:14]: is this used?

File: crates/http-api-bindings/src/embedding/llama.rs [hunk 3]
```diff
@@ -11,7 +11,7 @@ pub struct LlamaCppEngine {
 }
 
 impl LlamaCppEngine {
-    pub fn create(api_endpoint: &str, api_key: Option<String>) -> Self {
+    pub fn create(api_endpoint: &str, api_key: Option<String>) -> impl Embedding {
```
Comment (wsxiaoys) [inline crates/http-api-bindings/src/embedding/llama.rs:14]: revert

File: crates/http-api-bindings/src/embedding/mod.rs [hunk 4]
```diff
@@ -1,18 +1,31 @@
...
 pub async fn create(config: &HttpModelConfig) -> Arc<dyn Embedding> {
-    match config.kind.as_str() {
+    let rpm = if let Some(limit) = &config.request_limit {
+        limit.request_per_minute
+    } else {
+        debug!(
+            "No request limit specified for model {}, defaulting to 600 rpm",
+            config.kind
+        );
+        600
```
Comment (wsxiaoys) [inline crates/http-api-bindings/src/embedding/mod.rs:25]: make this a default option in `config.rs`:

example: https://demo.tabbyml.com/files/github/TabbyML/tabby/-/blob/f2cf549ad9ef01338ce30626095108d0ca1bf5ed/crates/tabby-common/src/config.rs#L243

File: crates/tabby-common/src/config.rs [hunk 5]
```diff
@@ -309,6 +312,12 @@ pub struct HttpModelConfig {
     pub additional_stop_words: Option<Vec<String>>,
 }
 
+#[derive(Serialize, Deserialize, Builder, Debug, Clone)]
+pub struct RequestLimit {
```
Comment (wsxiaoys) [inline crates/tabby-common/src/config.rs:316]:
```suggestion
pub struct RateLimit {
```

File: crates/tabby-common/src/config.rs [hunk 6]
```diff
@@ -289,6 +289,9 @@ pub struct HttpModelConfig {
     #[builder(default)]
     pub api_key: Option<String>,
 
+    #[builder(default)]
+    pub request_limit: Option<RequestLimit>,
```
Comment (wsxiaoys) [inline crates/tabby-common/src/config.rs:293]:
```suggestion
    pub rate_limit: Option<RequestLimit>,
```

**Review by wsxiaoys — COMMENTED** [review]
Otherwise LGTM

File: crates/http-api-bindings/src/embedding/rate_limit.rs [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+};
+
+use async_trait::async_trait;
+use futures::future::BoxFuture;
+use tabby_inference::Embedding;
+use tokio::sync::Mutex;
+use tower::{Service, ServiceBuilder, ServiceExt};
+
+struct EmbeddingService {
+    embedding: Arc<dyn Embedding>,
```
Comment (wsxiaoys) [inline crates/http-api-bindings/src/embedding/rate_limit.rs:14]: can this being `embedding: T` / `Box<dyn Embedding>`

Comment (zwpaper) [reply crates/http-api-bindings/src/embedding/rate_limit.rs:14]: will follow up in next release

**Pull request closed without merging**